File: frontend/lib/auth-access.ts
export type AccessStatus = "active" | "pending" | "blocked" | "unknown"

export interface AccessRequestInput {
  email: string
  situation_type?: string
  deadline_urgency?: string
  jurisdiction?: string
  note?: string
}

export interface AccessRequestResponse {
  ok: boolean
  status: string
  message: string
}

export interface InviteLookupResponse {
  found: boolean
  status: string
  email?: string | null
  situation_type?: string | null
  deadline_urgency?: string | null
  jurisdiction?: string | null
  expires_at?: string | null
  accepted_at?: string | null
}

export interface AuthMeResponse {
  authenticated: boolean
  access_status: AccessStatus
  subject?: string | null
  email?: string | null
  name?: string | null
  roles: string[]
  is_admin: boolean
  profile?: {
    subject: string
    email?: string | null
    name?: string | null
    status: AccessStatus
    situation_type?: string | null
    deadline_urgency?: string | null
    jurisdiction?: string | null
    first_matter_id?: string | null
    onboarding_completed_at?: string | null
  } | null
}

export interface BetaInvite {
  invite_id: string
  email?: string | null
  status: string
  roles: string[]
  situation_type?: string | null
  deadline_urgency?: string | null
  jurisdiction?: string | null
  created_at: string
  updated_at: string
  expires_at: string
  accepted_at?: string | null
  revoked_at?: string | null
}

export interface InviteRequest {
  request_id: string
  email: string
  status: string
  situation_type?: string | null
  deadline_urgency?: string | null
  jurisdiction?: string | null
  note?: string | null
  fulfilled_invite_id?: string | null
  created_at: string
  updated_at: string
}

export interface CreateInviteInput {
  email?: string
  roles?: string[]
  situation_type?: string
  deadline_urgency?: string
  jurisdiction?: string
  expires_in_days?: number
}

export interface CreateInviteResponse {
  invite: BetaInvite
  token: string
  invite_url_path: string
}

export async function requestAccess(input: AccessRequestInput): Promise<AccessRequestResponse> {
  return fetchJson("/api/ors/auth/access-request", {
    method: "POST",
    body: JSON.stringify(input),
  })
}

export async function lookupInvite(token: string): Promise<InviteLookupResponse> {
  return fetchJson(`/api/ors/auth/invites/${encodeURIComponent(token)}`)
}

export async function acceptInvite(token: string): Promise<AuthMeResponse> {
  return fetchJson(`/api/ors/auth/invites/${encodeURIComponent(token)}/accept`, {
    method: "POST",
  })
}

export async function listAccessRequests(): Promise<InviteRequest[]> {
  return fetchJson("/api/ors/admin/auth/access-requests")
}

export async function listInvites(): Promise<BetaInvite[]> {
  return fetchJson("/api/ors/admin/auth/invites")
}

export async function createInvite(input: CreateInviteInput): Promise<CreateInviteResponse> {
  return fetchJson("/api/ors/admin/auth/invites", {
    method: "POST",
    body: JSON.stringify(input),
  })
}

export async function revokeInvite(inviteId: string): Promise<BetaInvite> {
  return fetchJson(`/api/ors/admin/auth/invites/${encodeURIComponent(inviteId)}/revoke`, {
    method: "POST",
  })
}

async function fetchJson<T>(url: string, options: RequestInit = {}): Promise<T> {
  const headers = new Headers(options.headers)
  if (!headers.has("Content-Type") && typeof options.body === "string") {
    headers.set("Content-Type", "application/json")
  }
  const response = await fetch(url, {
    cache: "no-store",
    ...options,
    headers,
  })
  const body = await response.json().catch(() => ({}))
  if (!response.ok) {
    throw new Error(typeof body.error === "string" ? body.error : `Request failed: ${response.status}`)
  }
  return body as T
}

File: crates/orsgraph-api/src/models/auth_access.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct UserProfile {
    pub subject: String,
    pub email: Option<String>,
    pub name: Option<String>,
    pub status: String,
    #[serde(default)]
    pub roles: Vec<String>,
    pub situation_type: Option<String>,
    pub deadline_urgency: Option<String>,
    pub jurisdiction: Option<String>,
    pub first_matter_id: Option<String>,
    pub onboarding_completed_at: Option<String>,
    pub invite_id: Option<String>,
    pub created_at: String,
    pub updated_at: String,
    pub last_seen_at: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct BetaInvite {
    pub invite_id: String,
    pub token_hash: String,
    pub email: Option<String>,
    pub status: String,
    #[serde(default)]
    pub roles: Vec<String>,
    pub situation_type: Option<String>,
    pub deadline_urgency: Option<String>,
    pub jurisdiction: Option<String>,
    pub created_by_subject: Option<String>,
    pub accepted_by_subject: Option<String>,
    pub created_at: String,
    pub updated_at: String,
    pub expires_at: String,
    pub accepted_at: Option<String>,
    pub revoked_at: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct InviteRequest {
    pub request_id: String,
    pub email: String,
    pub status: String,
    pub situation_type: Option<String>,
    pub deadline_urgency: Option<String>,
    pub jurisdiction: Option<String>,
    pub note: Option<String>,
    pub fulfilled_invite_id: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Deserialize)]
pub struct CreateAccessRequest {
    pub email: String,
    pub situation_type: Option<String>,
    pub deadline_urgency: Option<String>,
    pub jurisdiction: Option<String>,
    pub note: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct CreateAccessRequestResponse {
    pub ok: bool,
    pub status: String,
    pub message: String,
}

#[derive(Debug, Serialize)]
pub struct InviteLookupResponse {
    pub found: bool,
    pub status: String,
    pub email: Option<String>,
    pub situation_type: Option<String>,
    pub deadline_urgency: Option<String>,
    pub jurisdiction: Option<String>,
    pub expires_at: Option<String>,
    pub accepted_at: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct AuthMeResponse {
    pub authenticated: bool,
    pub access_status: String,
    pub subject: Option<String>,
    pub email: Option<String>,
    pub name: Option<String>,
    #[serde(default)]
    pub roles: Vec<String>,
    pub is_admin: bool,
    pub profile: Option<UserProfile>,
}

#[derive(Debug, Deserialize)]
pub struct CreateInviteRequest {
    pub email: Option<String>,
    pub roles: Option<Vec<String>>,
    pub situation_type: Option<String>,
    pub deadline_urgency: Option<String>,
    pub jurisdiction: Option<String>,
    pub expires_in_days: Option<u64>,
}

#[derive(Debug, Serialize)]
pub struct CreateInviteResponse {
    pub invite: BetaInvite,
    pub token: String,
    pub invite_url_path: String,
}

#[derive(Debug, Deserialize)]
pub struct ListAuthQuery {
    pub status: Option<String>,
    pub limit: Option<usize>,
}

#[derive(Debug, Deserialize)]
pub struct PatchUserAccessRequest {
    pub status: Option<String>,
    pub roles: Option<Vec<String>>,
}
